cli/azd/cmd/up.go

```diff
@@ -91,15 +99,17 @@ func (u *upAction) Run(ctx context.Context) (*actions.ActionResult, error) {
 		return nil, fmt.Errorf("running init: %w", err)
 	}
 
-	_, err = u.infraCreate.Run(ctx)
+	provisionOptions := &middleware.Options{Name: "infracreate", Aliases: []string{"provision"}}
+	_, err = u.runner.RunAction(ctx, provisionOptions, u.infraCreate)
```

Comment:
The implication here is interesting if we're going with the route of middlewares in the inner actions. The `debug` middleware would fire multiple times, as does the `telemetry` which isn't the current behavior.

Reply:
> as does the telemetry which isn't the current behavior.

I think this new behavior (for telemetry) is what we actually want, right?  This ensures that we'll see child spans for each of the composite parts of the action tied to the parent span, which feels like the right thing long term right?

Reply:
Agree with @ellismg here that this would be good data to have so we can see the nested spans for composite operations.

Reply:
@ellismg @wbreza 

For telemetry, I think we'd want an action-level span instead of a silent byproduct of middleware invocation. While the current change does work for tracking perf purposes, it muddies the query for obtaining true command-usage, i.e., the entry-point. We want to easily measure how many times a user runs an actual command name. With this behavior, we'd emit a command-level event `cmd.<name>` per action which is slightly incorrect.

For debug middleware, action-level invocation is incorrect. Fundamentall, we want to be consistent on whether we want command-level middleware or action-level middleware. If we want action-level middleware, there's a different way to accomplish this that would be solidifed in the action design.

Reply:
> Fundamentally, we want to be consistent on whether we want command-level middleware or action-level middleware (or both). If we want action-level middleware, there's a different way to accomplish this that would be solidifed in the action design.

I agree this is the crux of the issue (and a great way of stating it).  It does seem like the fundamental problem here is that the debug middleware wants to be a "command level middleware" and for telemetry we want both "command level middleware" (so we can emit the `cmd.<command-name>` spans, and perhaps add attributes like "return code") as well as action level middleware (so we can have things like "action.infra-create" which is emitted as a sub span of both `cmd.infra-create` and `cmd.up`)

> And if we can exclude telemetry from this behavior until we really think through it -- I'd be okay with it.

I'd also be fine with pulling this out for now if it makes it easier for us to land the overall infrastructure.  Since presently this middleware is "action" level middleware, I guess the other option would be to change the names of the spans emitted by it (to use the action prefix) and then in our common .RunE implementation that we synthesize register the `cmd` span?

> Hope I'm not too pedantic here...

Not at all.  I really appreciate it.

Reply:
The middleware runner now sets a flag when it execs a sub action.  Middleware components can now inspect this flag and do the right thing.  For the case of both telemetry & debug middleware I have gone ahead and short circuit when it detects a sub action is executing.

This enhancement has also allowed us to bring back the UX middleware where it will evaluate the action result only at the root/top level command.  This will allow for a cleaner separation of concerns since the command builder/runner shouldn't need to be concerned with anything UX related.